### PR TITLE
[WIP] Temporary Android branch build

### PR DESCRIPTION
**Pull request by themarpe**
Currently targets a temporary development branch of DepthAI with Android support.
Afterwards, when Android support is mainlined the changes will be able to be merged

### CODE REVIEW

**Comment by saudet** [comment]
BTW, instead of coming up with custom toolchains, try to use use what's in CMake since 3.21:
https://developer.android.com/ndk/guides/cmake

**Comment by themarpe** [comment]
@saudet 
I guess the CI not triggering because it was based off an older commit / had conflicts 
`6f8889c3f3a7b91fe55d96195dd976b643da68f1` ( * Upgrade presets for Tesseract 5.0.0, DepthAI 2.13.3)

I've based it of and basically did a revert to target older depthai version. (and seems like it needs approval again, probably because of the force push - not sure / or was it always requiring approval, per commit)

**Comment by saudet** [comment]
BTW, the workflows are already running from your fork, you don't need to make them work here:
https://github.com/luxonis/javacpp-presets/actions

**Comment by saudet** [comment]
And the reason why we need to approve commits from external contributors is for security reasons:
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
This isn't needed when you make contributions to forks you own, of course.

**Comment by themarpe** [comment]
Okay, seems like I got the build working. (https://github.com/luxonis/javacpp-presets/actions/runs/1696048933)
Where should I look for Android prebuilts? (or are those are only uploaded when CI runs on your end?)

**Comment by saudet** [comment]
For security reasons again, the binaries from PR builds don't get uploaded anywhere. You can however upload the ones built from your fork anywhere you want, including Sonatype OSSRH: https://central.sonatype.org/publish/publish-guide/

That said, I could merge your PR in this repository and it would start uploading snapshots that way with my credentials. Now, can you rebase this on DepthAI 2.14.x? Or is this going to stay on 2.13.x for a while? If it's the latter, I'll create a special branch where to merge stuff for now until it's ready to be merged to the main branch.

**Comment by themarpe** [comment]
All makes sense. Though that these get uploaded as "temporary builds" somewhere. 

Hmm, i dont want to pollute the reguar versioning - right now I added the comit after the version, would that do to create builds that arent version only?

Otherwise it might be better if I upload them manually somewhere... How can I get out the built artifacts? (Or I see that I might be able to just supply the relevant secrets and rerun? I assume usign the above mentioned Sonatype repo. Would that be the simplest repo to get some snapshots uploaded?)

 Yeah, regarding versioning that was the main issue - most likely won't be mainlined soon (I could expedite it a bit, bot it still requires a bit work around everything and testing...), so wanted just to create prebuilts based off this special branch that supports Android.

**Comment by themarpe** [comment]
I checked and it seems that Artifactory has a Maven repository option - I could spin a repo there easily - would credentials to that be adequate to get the builds deployed to it?

**Comment by saudet** [comment]
> Hmm, i dont want to pollute the reguar versioning - right now I added the comit after the version, would that do to create builds that arent version only?

You can give them a different version, by appending the commit hash or whatever, sure, that's fine.

> I checked and it seems that Artifactory has a Maven repository option - I could spin a repo there easily - would credentials to that be adequate to get the builds deployed to it?

You can do whatever you want from your fork, but if you're not already using something somewhere, the easiest option is probably to stay here on GitHub with GitHub Packages: https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-apache-maven-registry

**Comment by saudet** [comment]
In any case, I've created a branch named "depthai" that is set to the last commit before upgrading to 2.14.0, and rebased this PR on it. We can merge this there anytime you're ready, but please first revert unrelated changes. Thanks!

**Comment by themarpe** [comment]
@saudet thanks a lot, I think this will be the easiest:)

I've force-pushed the changes, which should now be based on `depthai` branch.

File: .github/workflows/depthai.yml
Comment (saudet) [inline .github/workflows/depthai.yml:62]: Please put everything for Android *before* the other platforms, in this file, cppbuild.sh, and pom.xml files as well.

File: depthai/README.md
Comment (saudet) [inline depthai/README.md:4]: Please also remove extra blank spaces added here and in a few other places.

File: pom.xml
Comment (saudet) [inline pom.xml:768]: Please reinsert it in the same order as the platforms.

**Review by saudet — CHANGES_REQUESTED** [review]
There's also missing entries for Android in these files:
https://github.com/luxonis/javacpp-presets/blob/depthai-android/depthai/platform/pom.xml
https://github.com/luxonis/javacpp-presets/blob/depthai-android/depthai/src/main/java/org/bytedeco/depthai/presets/depthai.java#L47

Normally I'd fix these nits myself, but by default we don't get permission to make changes to pull requests from repositories of organizations, only when it's from a personal repo:
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork
Sorry for the trouble!

**Comment by themarpe** [comment]
No issue @saudet - thanks so much for helping out with this!

I've force pushed again - let me know if I missed something or should modify something else:)

File: pom.xml
Comment (saudet) [inline pom.xml:350]: Let's also revert this, even if it shouldn't be there, to help with any merge conflicts later...

File: depthai/src/main/java/org/bytedeco/depthai/presets/depthai.java
Comment (saudet) [inline depthai/src/main/java/org/bytedeco/depthai/presets/depthai.java:328]: BTW, `dai::DeviceInfo::toString()` doesn't exist in main branch. Is something new in the branch for Android?

File: .github/workflows/depthai.yml
Comment (saudet) [inline .github/workflows/depthai.yml:64]: Still need to change the order...

File: depthai/cppbuild.sh
Comment (saudet) [inline depthai/cppbuild.sh:107]: ... and here ...

**Review by saudet — CHANGES_REQUESTED** [review]
And missing entries here:

File: depthai/src/main/java/org/bytedeco/depthai/presets/depthai.java
Comment (themarpe) [reply depthai/src/main/java/org/bytedeco/depthai/presets/depthai.java:328]: Yeah, some improvements on that branch, but they got left behind for now, until mainlined

Comment (themarpe) [reply depthai/src/main/java/org/bytedeco/depthai/presets/depthai.java:328]: Added because it was failing otherwise with an overload between DeviceInfo toString and base class Pointer...

**Comment by themarpe** [comment]
Force pushed - my bad about whitespace changes. I have diffs for those disabled and VSCode set to automatically trim trialing whitespace, so it kinda slips past my local review

**Comment by themarpe** [comment]
Android not building anymore ... Doesn't seem to detect OpenCV library anymore, which causes it to fail later on.
Any ideas? (the differences were adding android entries to `depthai/platform/depthai.xml` and to preset `depthai.java`, those I think might be relevant)

**Comment by saudet** [comment]
It looks like a problem with the build options. It says OpenCV support isn't enabled:
```
/__w/javacpp-presets/javacpp-presets/depthai/cppbuild/android-arm/include/depthai/pipeline/datatype/ImgFrame.hpp:189:9:
error: static_assert failed due to requirement 'dependent_false<bool>::value' "Library not configured with OpenCV support"
```

**Comment by saudet** [comment]
OpenCV installs its CMake-related files a bit differently for Android.
You'll probably need to set its directory somewhere, for example, see:
https://github.com/bytedeco/javacpp-presets/blob/master/flandmark/cppbuild.sh

**Comment by themarpe** [comment]
Thanks @saudet - fixed one extra issue on DepthAI side, but now the CI is passing fully:)

**Comment by saudet** [comment]
Ok, Android builds are up here:
https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/depthai/
https://oss.sonatype.org/content/repositories/snapshots/org/bytedeco/depthai-platform/
You can use them in your Gradle builds as explained here: http://bytedeco.org/builds/

/cc @Grumpy141 @wureka

**Comment by themarpe** [comment]
Thanks so much for leading me through this, and deploying this on your end:)

Looking forward to bringing Android to mainline, so this special branch can be merged as well:)

A couple of questions:
 - How should I update the library if there is something that needs to be fixed?
 - Any ideas on the following:
 ```
extern "C" JNIEXPORT jstring JNICALL
Java_com_luxonis_depthaiexampleapp_MainActivity_initXLink(
        JNIEnv* env,
        jobject /* this */) {

    JavaVM* jvm;
    env->GetJavaVM(&jvm);
    // ... use the JavaVM pointer to setup libusb underneath
``` 

Where would be the cleanest place to execute this kind of code? For libusb to work on Android, it requires getting JavaENV / JavaVM pointer so it can scan for devices and ask for permissions (from Android SDK).

I assume that'll be between depthai-core and javacpp - I saw a jnidepthai library/file? being created in the build, which has the required JNI code I assume. Thats a spot I think I'd be able to get the pointer from and call the init function.
Thoughts?

**Comment by saudet** [comment]
Just keep sending pull requests against this branch to have updates merged and deployed!

For the init function, this looks like something that should be kept in depth-core, but if you export a JNI method, you'll need to provide a Java interface and everything to initialize that, which JavaCPP already does, so I'd just make it available as a normal C function, and then we could easily call it from Java with JavaCPP like this for FFmpeg:
https://github.com/bytedeco/javacv/blob/master/src/main/java/org/bytedeco/javacv/FFmpegFrameGrabber.java#L117

**Comment by saudet** [comment]
Besides, there are a lot of developers for Android that do not code in Java, so having an init method that is only available from Java isn't something you want to do.

**Comment by themarpe** [comment]
This is perfect! Just what I was looking for. Thanks!

**Comment by themarpe** [comment]
I agree, just added the changes to `depthai-core` and opened the PR: https://github.com/bytedeco/javacpp-presets/pull/1126

**Pull request merged**
